zip-0315.rst

```diff
@@ -1,9 +1,629 @@
+    most preferred (usually the most recent) pool on receipt.
+
+Opportunistic migration
+    The process of transferring previously received shielded funds from older pools
+    to the most preferred (usually the most recent) pool as part of a user-initiated
+    transaction.
+
+Transaction output (TXO)
+    An output (transparent coin or shielded note) of a transaction on the consensus
+    block chain or in the mempool visible to a wallet.
+
+
+Motivation
+==========
+
+Zcash wallets have to serve two purposes as a user agent:
+
+* to manage the economic actions dictated by the user;
+* to manage the privacy implications of these actions.
+
+The goal of this ZIP is to document security and privacy best practices when handling
+funds and transactions as the user's agent. These best practices are intended to
+provide as much privacy as is feasible by default, while still enabling the user's
+desired transactions to occur, and exposing any privacy implications to the user so
+that they have enough information to assess the consequences of their economic actions.
+
+This ZIP covers best practices for:
+
+* what information to display to the user about transactions and balances;
+* how to handle interactions between the ZIP 32 [#zip-0032]_ key tree and Unified Addresses [#zip-0316]_;
+* when to use external or internal keys/addresses;
+* sharing addresses and viewing keys;
+* sending and receiving funds (including fee selection);
+* migrating funds between pools.
+
+
+Requirements
+============
+
+User consent
+------------
+
+A wallet typically reveals some information in the process of creating a transaction.
+Which information is revealed depends on the configured wallet privacy policy.
+The guiding principle of these requirements is that users must explicitly consent
+to each instance of information being revealed by the wallet in a transaction.
+
+A user may give blanket consent to reveal a particular kind of information, and
+must also be able to change the configured wallet privacy policy to avoid the
+wallet creating new information leaks of a given type.
+
+The specifications below describe some situations in which blanket consent may be
+inappropriate.
+
+Some varieties of consent may not be revocable, for example if a user chooses to
+share some of their keys.
+
+
+Prompt accessibility of funds
+-----------------------------
+
+Wallets need to take account of two concerns:
+
+* enabling funds to be spent as quickly as possible to reduce latency;
+* waiting long enough before spending TXOs to ensure that the
+  confirmed-spendable balance is not overestimated, and so can be trusted
+  by the user.
+
+To enable this we define two kinds of TXOs:
+
+* A trusted TXO is a TXO received from a party where the wallet trusts a
+  double-spend not to occur, e.g. TXOs created by the wallet's internal TXO
+  handling.
+* An untrusted TXO is a TXO received by the wallet that is not trusted.
+
```

Comment:
This puts an undue emphasis on double-spends. It's possible for a shielded output to become unmined (and not remined) without a double-spend, because its anchor has been invalidated.

```suggestion
* A trusted TXO is a TXO received from a party where the wallet trusts that
  it will remain mined in its original transaction, such as TXOs created by
  the wallet's internal TXO handling.
* An untrusted TXO is a TXO received by the wallet that is not trusted.
```